bin/agent-data-plane/src/internal/remote_agent.rs

```diff
@@ -177,68 +285,114 @@ impl RemoteAgentImpl {
+            resp
+        })
+    }
 }
 
+// Implement the three separate service traits instead of the old unified RemoteAgent trait
+
 #[async_trait]
-impl RemoteAgent for RemoteAgentImpl {
+impl StatusProvider for RemoteAgentImpl {
     async fn get_status_details(
         &self, _request: tonic::Request<GetStatusDetailsRequest>,
     ) -> Result<tonic::Response<GetStatusDetailsResponse>, tonic::Status> {
-        let app_details = saluki_metadata::get_app_details();
-
-        let mut builder = StatusBuilder::new();
-        builder
-            .main_section()
-            .set_field("Version", app_details.version().raw())
-            .set_field("Git Commit", app_details.git_hash())
-            .set_field("Architecture", app_details.target_arch())
-            .set_field("Started", self.started.to_rfc3339());
-
-        self.write_dsd_metrics(&mut builder);
-
-        Ok(tonic::Response::new(builder.into_response()))
-    }
-
-    async fn get_flare_files(
-        &self, _request: tonic::Request<GetFlareFilesRequest>,
-    ) -> Result<tonic::Response<GetFlareFilesResponse>, tonic::Status> {
-        let response = GetFlareFilesResponse {
-            files: HashMap::default(),
-        };
-        Ok(tonic::Response::new(response))
+        return self
+            .session_id_middleware(async || {
+                let app_details = saluki_metadata::get_app_details();
+
+                let mut builder = StatusBuilder::new();
+                builder
+                    .main_section()
+                    .set_field("Version", app_details.version().raw())
+                    .set_field("Git Commit", app_details.git_hash())
+                    .set_field("Architecture", app_details.target_arch())
+                    .set_field("Started", self.started.to_rfc3339());
+
+                self.write_dsd_metrics(&mut builder);
+
+                Ok(tonic::Response::new(builder.into_response()))
+            })
+            .await;
     }
+}
 
+#[async_trait]
+impl TelemetryProvider for RemoteAgentImpl {
     async fn get_telemetry(
         &self, _request: tonic::Request<GetTelemetryRequest>,
     ) -> Result<tonic::Response<GetTelemetryResponse>, tonic::Status> {
-        // Telemetry is not enabled.
-        if self.prometheus_listen_addr.is_none() {
-            return Ok(tonic::Response::new(GetTelemetryResponse { payload: None }));
-        }
-
-        let prometheus_listen_addr = self.prometheus_listen_addr.unwrap();
-        let mut client: HttpClient<String> = HttpClient::builder().build().unwrap();
-
-        let uri_string = format!("http://{}", prometheus_listen_addr);
-        let uri: Uri = uri_string.parse().unwrap();
-        let request = Request::builder()
-            .uri(uri)
-            .body(String::new())
-            .map_err(|e| tonic::Status::internal(e.to_string()))?;
-
-        let resp = client.send(request).await.unwrap();
+        return self
+            .session_id_middleware(async || {
```

Comment:
I'm not in _love_ with this pattern of each gRPC service method needs to wrap their logic in an async closure to get the session ID attached to the response.

I'd rather see this be attached, uniformly, at the point where we're adding the services to the API builder.. but that would require some more code here which isn't relevant to the functional behavior and I'm not _exactly_ sure how I would go about that, so I'm only saying this in a non-blocking way so that I don't forget about it. 😅